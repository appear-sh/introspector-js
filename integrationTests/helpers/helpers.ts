import { spawn } from "child_process"

export async function startFrameworkServer(
  framework: string,
  collectorUrl: string,
): Promise<{ port: number; stop: () => Promise<void> }> {
  return new Promise((resolve, reject) => {
    const server = spawn(
      "node",
      [
        "--import",
        "./integrationTests/helpers/hook.ts",
        "--loader",
        "ts-node/esm",
        `integrationTests/frameworks/${framework.toLowerCase()}.ts`,
      ],
      {
        env: {
          ...process.env,
          COLLECTOR_URL: collectorUrl,
        },
        stdio: ["pipe", "pipe", "pipe"],
      },
    )

    let port: number | undefined
    let error = ""

    server.stdout.on("data", (data) => {
      const output = data.toString()
      console.log(`[${framework}] ${output}`)

      // Look for the port number in the output
      const portMatch = output.match(/Server started on port (\d+)/)
      if (portMatch) {
        port = parseInt(portMatch[1], 10)
        resolve({
          port,
          stop: () =>
            new Promise<void>((resolve) => {
              server.kill()
              server.on("close", () => resolve())
            }),
        })
      }
    })

    server.stderr.on("data", (data) => {
      error += data.toString()
      console.error(`[${framework}] ${data}`)
    })

    server.on("error", (err) => {
      reject(
        new Error(
          `Failed to start ${framework} server: ${err.message}\n${error}`,
        ),
      )
    })

    server.on("close", (code) => {
      if (!port) {
        reject(
          new Error(
            `Server closed before starting. Exit code: ${code}\n${error}`,
          ),
        )
      }
    })
  })
}
