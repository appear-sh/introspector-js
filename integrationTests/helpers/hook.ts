import { registerAppear } from '../../src/otel/index.js';

// Get collector URL from environment
const collectorUrl = process.env.COLLECTOR_URL;
if (!collectorUrl) {
  throw new Error('COLLECTOR_URL environment variable is required');
}

// Register OpenTelemetry with mock collector
registerAppear({
  apiKey: 'test-key',
  environment: 'test',
  reporting: {
    endpoint: collectorUrl,
    batchIntervalSeconds: 0, // Disable batching for tests
  }
}); 