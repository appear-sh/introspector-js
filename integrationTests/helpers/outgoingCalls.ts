export async function makeHttpRequest() {
  return new Promise(async (resolve, reject) => {
    const https = await import("node:https")
    const httpRequest = https.request(
      "https://httpbin.org/anything",
      { method: "POST", headers: { "Content-Type": "application/json" } },
      (httpResponse) => {
        let data = ""
        httpResponse.on("data", (chunk) => {
          data += chunk
        })
        httpResponse.on("end", () => resolve(JSON.parse(data)))
      },
    )
    httpRequest.on("error", (error) => reject(error))
    httpRequest.write(JSON.stringify({ "http.request": "http.request" }))
    httpRequest.end()
  })
}

export async function makeFetchRequest() {
  return fetch("https://httpbin.org/anything", {
    method: "POST",
    body: JSON.stringify({ fetch: "fetch" }),
    headers: { "Content-Type": "application/json" },
  })
}

export async function makeOutgoingCalls() {
  await Promise.all([makeHttpRequest(), makeFetchRequest()])
}
