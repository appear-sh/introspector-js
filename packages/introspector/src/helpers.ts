import { ResolvedAppearConfig } from "./config.js"

export const isNonNullable = <T extends any>(
  value: T,
): value is NonNullable<T> => {
  return typeof value !== "undefined" && value !== null
}

export const defaultInterceptFilter = (
  request: Request,
  response: Response,
  config: ResolvedAppearConfig,
): boolean => {
  // Fetch & XHR don't have set destination
  // Probably something we don't care about, ignore.
  if (request.destination !== "") return false
  // ignore reports to Appear
  if (request.url === config.reporting.endpoint) return false

  return true
}
