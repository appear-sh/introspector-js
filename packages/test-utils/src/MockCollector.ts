import { createServer } from "http"
import type { AddressInfo } from "net"
import { EventEmitter } from "events"

export class MockCollector {
  private server: any
  private port: number = 0
  private traces: any[] = []
  private emitter = new EventEmitter()

  async start() {
    return new Promise<void>((resolve) => {
      this.server = createServer((req, res) => {
        if (req.method === "POST") {
          let body = ""
          req.on("data", (chunk) => (body += chunk))
          req.on("end", () => {
            try {
              const trace = JSON.parse(body)
              this.traces.push(trace)
              this.emitter.emit("trace", trace)
              res.writeHead(200)
              res.end()
            } catch (error) {
              console.error("Error parsing trace:", error)
              res.writeHead(500)
              res.end()
            }
          })
        } else {
          res.writeHead(405)
          res.end()
        }
      })

      this.server.listen(0, () => {
        this.port = (this.server.address() as AddressInfo).port
        resolve()
      })
    })
  }

  async stop() {
    return new Promise<void>((resolve) => {
      this.server.close(() => resolve())
    })
  }

  getUrl() {
    return `http://localhost:${this.port}`
  }

  getTraces() {
    return this.traces
  }

  waitForTrace(timeout = 5000): Promise<any> {
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        reject(new Error("Timeout waiting for trace"))
      }, timeout)

      this.emitter.once("trace", (trace) => {
        clearTimeout(timer)
        resolve(trace)
      })
    })
  }
}
