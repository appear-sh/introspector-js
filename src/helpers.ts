import { IncomingMessage, ServerResponse } from "http"
import { EventEmitter } from "events"

import {
  APPEAR_SYMBOL,
  INTROSPECTOR_EMITTER_SYMBOL,
  INTROSPECTOR_HOOKED_SYMBOL,
} from "./hook/symbol"
import { GlobalAppear } from "./global"

export const isNonNullable = <T extends any>(
  value: T,
): value is NonNullable<T> => {
  return typeof value !== "undefined" && value !== null
}

export async function readBodyFromRequest(
  req: IncomingMessage,
): Promise<string> {
  return new Promise((resolve) => {
    const chunks: Buffer[] = []
    req.on("data", (chunk: Buffer) => chunks.push(chunk))
    req.on("end", () => {
      const body = Buffer.concat(chunks).toString()
      resolve(body)
    })
  })
}

export function readBodyFromResponse(res: ServerResponse): Promise<string> {
  return new Promise((resolve) => {
    const originalWrite = res.write.bind(res)
    const originalEnd = res.end.bind(res)
    const chunks: Buffer[] = []

    res.write = function (
      chunk: any,
      encodingOrCallback?:
        | BufferEncoding
        | ((error: Error | null | undefined) => void),
      callback?: (error: Error | null | undefined) => void,
    ): boolean {
      const encoding = (encodingOrCallback as BufferEncoding) || "utf8"
      chunks.push(Buffer.from(chunk, encoding))
      return originalWrite.call(res, chunk, encoding, callback)
    }

    res.end = function (
      chunk: any,
      ...args: any[]
    ): ServerResponse<IncomingMessage> {
      if (chunk) {
        chunks.push(Buffer.from(chunk))
      }
      const responseBody = Buffer.concat(chunks).toString()

      resolve(responseBody)

      const encoding = args[0] || "utf8"
      return originalEnd.call(res, chunk, encoding, ...args.slice(1))
    }
  })
}

export async function incomingMessageToRequest(
  req: IncomingMessage,
): Promise<Request> {
  const protocol = (req.headers["x-forwarded-proto"] as string) || "http"
  const host = req.headers.host
  const url = `${protocol}://${host}${req.url}`

  const headers = new Headers(req.headers as Record<string, string>)

  let body = null

  if (req.method !== "GET" && req.method !== "HEAD") {
    body = await readBodyFromRequest(req)
  }

  return new Request(url, {
    method: req.method,
    headers: headers,
    body,
  })
}

export async function serverResponseToResponse(
  res: ServerResponse<IncomingMessage>,
): Promise<Response> {
  // MUST read the body first, as headers etc are likely sent before this.
  const body = await readBodyFromResponse(res)

  const headers = new Headers()
  const outgoingHeaders = res.getHeaders()
  for (const [headerName, headerValue] of Object.entries(outgoingHeaders)) {
    if (headerValue) {
      headers.append(
        headerName,
        Array.isArray(headerValue)
          ? headerValue.join(", ")
          : String(headerValue),
      )
    }
  }

  return new Response(body, {
    status: res.statusCode,
    statusText: res.statusMessage,
    headers,
  })
}

// Typescript doesn't seem to support using symbols as
// accessor keys for globalThis, so we need to cast to any.

export function setGlobalAppear(emitter: EventEmitter) {
  ;(globalThis as any)[APPEAR_SYMBOL] = {
    [INTROSPECTOR_EMITTER_SYMBOL]: emitter,
    [INTROSPECTOR_HOOKED_SYMBOL]: true,
  }
}

export function getGlobalAppear(): GlobalAppear {
  return (
    (globalThis as any)[APPEAR_SYMBOL] ?? {
      [INTROSPECTOR_EMITTER_SYMBOL]: new EventEmitter(),
      [INTROSPECTOR_HOOKED_SYMBOL]: false,
    }
  )
}
