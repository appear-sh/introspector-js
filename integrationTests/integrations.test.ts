import { describe, it, expect, beforeAll, afterAll } from "vitest"
import { MockCollector } from "./helpers/MockCollector.js"
import { startFrameworkServer } from "./helpers/helpers.js"
import stringify from "fast-json-stable-stringify"

// Test server implementations for different frameworks
const frameworks = [
  "express",
  "fastify",
  // Add more frameworks here (Next.js, etc.)
]

// Create test suite for each framework
frameworks.forEach((framework) => {
  describe(`Framework: ${framework}`, () => {
    let collector: MockCollector
    let server: { port: number; stop: () => Promise<void> }

    beforeAll(async () => {
      // Start mock collector
      collector = new MockCollector()
      await collector.start()

      // Start framework server in a separate process
      server = await startFrameworkServer(framework, collector.getUrl())
    })

    afterAll(async () => {
      await Promise.all([collector.stop(), server.stop()])
    })

    it(
      "should capture traces for incoming request and outgoing calls",
      { timeout: 30000 },
      async () => {
        // Make test request
        const response = await fetch(`http://localhost:${server.port}/test`, {
          method: "POST",
          headers: { "Content-Type": "application/json" },
          body: JSON.stringify({ test: "data" }),
        })

        expect(response.status).toBe(200)
        const data = await response.json()
        expect(data.message).toBe("Success")

        // Wait for and verify traces
        const trace = await collector.waitForTrace(30000)

        // Censor ports in URIs and sort operations for consistent snapshots
        const operations = trace.operations
          .map((op) => {
            op.request.uri = op.request.uri.replace(/:\d+/, ":PORT")
            return op
          })
          .toSorted((a, b) => stringify(a).localeCompare(stringify(b)))

        expect(operations).toMatchSnapshot()
        expect(operations).toHaveLength(3)
      },
    )
  })
})
