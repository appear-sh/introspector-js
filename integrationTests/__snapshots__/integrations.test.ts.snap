// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Framework: express > should capture traces for incoming request and outgoing calls 1`] = `
[
  {
    "direction": "incoming",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "test": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
          },
          "required": [
            "test",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 13,
          "minLength": 13,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 2,
          "minLength": 2,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "http://localhost:PORT/test",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "body": {
              "properties": {
                "test": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
              },
              "required": [
                "test",
              ],
              "type": "object",
            },
            "message": {
              "maxLength": 7,
              "minLength": 7,
              "type": "string",
            },
          },
          "required": [
            "message",
            "body",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "content-length": {
          "maxLength": 2,
          "minLength": 2,
          "type": "string",
        },
        "content-type": {
          "maxLength": 31,
          "minLength": 31,
          "type": "string",
        },
        "etag": {
          "maxLength": 34,
          "minLength": 34,
          "type": "string",
        },
        "x-powered-by": {
          "maxLength": 7,
          "minLength": 7,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "source": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
          },
          "required": [
            "source",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 11,
          "minLength": 11,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "args": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "data": {
              "maxLength": 25,
              "minLength": 25,
              "type": "string",
            },
            "files": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "form": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "headers": {
              "properties": {
                "Content-Length": {
                  "maxLength": 2,
                  "minLength": 2,
                  "type": "string",
                },
                "Content-Type": {
                  "maxLength": 16,
                  "minLength": 16,
                  "type": "string",
                },
                "Host": {
                  "format": "filename",
                  "type": "string",
                },
                "Traceparent": {
                  "maxLength": 55,
                  "minLength": 55,
                  "type": "string",
                },
                "X-Amzn-Trace-Id": {
                  "maxLength": 40,
                  "minLength": 40,
                  "type": "string",
                },
              },
              "required": [
                "Content-Length",
                "Content-Type",
                "Host",
                "Traceparent",
                "X-Amzn-Trace-Id",
              ],
              "type": "object",
            },
            "json": {
              "properties": {
                "source": {
                  "maxLength": 12,
                  "minLength": 12,
                  "type": "string",
                },
              },
              "required": [
                "source",
              ],
              "type": "object",
            },
            "method": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
            "origin": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
            "url": {
              "format": "uri",
              "type": "string",
            },
          },
          "required": [
            "args",
            "data",
            "files",
            "form",
            "headers",
            "json",
            "method",
            "origin",
            "url",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": null,
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 17,
          "minLength": 17,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "args": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "data": {
              "maxLength": 18,
              "minLength": 18,
              "type": "string",
            },
            "files": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "form": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "headers": {
              "properties": {
                "Accept": {
                  "maxLength": 3,
                  "minLength": 3,
                  "type": "string",
                },
                "Accept-Encoding": {
                  "maxLength": 17,
                  "minLength": 17,
                  "type": "string",
                },
                "Accept-Language": {
                  "maxLength": 1,
                  "minLength": 1,
                  "type": "string",
                },
                "Content-Length": {
                  "maxLength": 2,
                  "minLength": 2,
                  "type": "string",
                },
                "Content-Type": {
                  "maxLength": 16,
                  "minLength": 16,
                  "type": "string",
                },
                "Host": {
                  "format": "filename",
                  "type": "string",
                },
                "Sec-Fetch-Mode": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
                "Traceparent": {
                  "maxLength": 55,
                  "minLength": 55,
                  "type": "string",
                },
                "User-Agent": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
                "X-Amzn-Trace-Id": {
                  "maxLength": 40,
                  "minLength": 40,
                  "type": "string",
                },
              },
              "required": [
                "Accept",
                "Accept-Encoding",
                "Accept-Language",
                "Content-Length",
                "Content-Type",
                "Host",
                "Sec-Fetch-Mode",
                "Traceparent",
                "User-Agent",
                "X-Amzn-Trace-Id",
              ],
              "type": "object",
            },
            "json": {
              "properties": {
                "source": {
                  "maxLength": 5,
                  "minLength": 5,
                  "type": "string",
                },
              },
              "required": [
                "source",
              ],
              "type": "object",
            },
            "method": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
            "origin": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
            "url": {
              "format": "uri",
              "type": "string",
            },
          },
          "required": [
            "args",
            "data",
            "files",
            "form",
            "headers",
            "json",
            "method",
            "origin",
            "url",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
]
`;

exports[`Framework: fastify > should capture traces for incoming request and outgoing calls 1`] = `
[
  {
    "direction": "incoming",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "test": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
          },
          "required": [
            "test",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 13,
          "minLength": 13,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 2,
          "minLength": 2,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "http://localhost:PORT/test",
    },
    "response": {
      "body": {
        "contentMediaType": "text/plain",
        "type": "string",
      },
      "headers": {
        "content-type": {
          "maxLength": 24,
          "minLength": 24,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "source": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
          },
          "required": [
            "source",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "host": {
          "maxLength": 11,
          "minLength": 11,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "args": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "data": {
              "maxLength": 25,
              "minLength": 25,
              "type": "string",
            },
            "files": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "form": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "headers": {
              "properties": {
                "Content-Length": {
                  "maxLength": 2,
                  "minLength": 2,
                  "type": "string",
                },
                "Content-Type": {
                  "maxLength": 16,
                  "minLength": 16,
                  "type": "string",
                },
                "Host": {
                  "format": "filename",
                  "type": "string",
                },
                "Traceparent": {
                  "maxLength": 55,
                  "minLength": 55,
                  "type": "string",
                },
                "X-Amzn-Trace-Id": {
                  "maxLength": 40,
                  "minLength": 40,
                  "type": "string",
                },
              },
              "required": [
                "Content-Length",
                "Content-Type",
                "Host",
                "Traceparent",
                "X-Amzn-Trace-Id",
              ],
              "type": "object",
            },
            "json": {
              "properties": {
                "source": {
                  "maxLength": 12,
                  "minLength": 12,
                  "type": "string",
                },
              },
              "required": [
                "source",
              ],
              "type": "object",
            },
            "method": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
            "origin": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
            "url": {
              "format": "uri",
              "type": "string",
            },
          },
          "required": [
            "args",
            "data",
            "files",
            "form",
            "headers",
            "json",
            "method",
            "origin",
            "url",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
  {
    "direction": "outgoing",
    "request": {
      "body": null,
      "headers": {
        "accept": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "accept-encoding": {
          "maxLength": 17,
          "minLength": 17,
          "type": "string",
        },
        "accept-language": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "sec-fetch-mode": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
        "traceparent": {
          "maxLength": 55,
          "minLength": 55,
          "type": "string",
        },
        "user-agent": {
          "maxLength": 4,
          "minLength": 4,
          "type": "string",
        },
      },
      "method": "POST",
      "query": {},
      "uri": "https://httpbin.org/anything",
    },
    "response": {
      "body": {
        "contentMediaType": "application/json",
        "contentSchema": {
          "properties": {
            "args": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "data": {
              "maxLength": 18,
              "minLength": 18,
              "type": "string",
            },
            "files": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "form": {
              "properties": {},
              "required": [],
              "type": "object",
            },
            "headers": {
              "properties": {
                "Accept": {
                  "maxLength": 3,
                  "minLength": 3,
                  "type": "string",
                },
                "Accept-Encoding": {
                  "maxLength": 17,
                  "minLength": 17,
                  "type": "string",
                },
                "Accept-Language": {
                  "maxLength": 1,
                  "minLength": 1,
                  "type": "string",
                },
                "Content-Length": {
                  "maxLength": 2,
                  "minLength": 2,
                  "type": "string",
                },
                "Content-Type": {
                  "maxLength": 16,
                  "minLength": 16,
                  "type": "string",
                },
                "Host": {
                  "format": "filename",
                  "type": "string",
                },
                "Sec-Fetch-Mode": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
                "Traceparent": {
                  "maxLength": 55,
                  "minLength": 55,
                  "type": "string",
                },
                "User-Agent": {
                  "maxLength": 4,
                  "minLength": 4,
                  "type": "string",
                },
                "X-Amzn-Trace-Id": {
                  "maxLength": 40,
                  "minLength": 40,
                  "type": "string",
                },
              },
              "required": [
                "Accept",
                "Accept-Encoding",
                "Accept-Language",
                "Content-Length",
                "Content-Type",
                "Host",
                "Sec-Fetch-Mode",
                "Traceparent",
                "User-Agent",
                "X-Amzn-Trace-Id",
              ],
              "type": "object",
            },
            "json": {
              "properties": {
                "source": {
                  "maxLength": 5,
                  "minLength": 5,
                  "type": "string",
                },
              },
              "required": [
                "source",
              ],
              "type": "object",
            },
            "method": {
              "maxLength": 4,
              "minLength": 4,
              "type": "string",
            },
            "origin": {
              "maxLength": 12,
              "minLength": 12,
              "type": "string",
            },
            "url": {
              "format": "uri",
              "type": "string",
            },
          },
          "required": [
            "args",
            "data",
            "files",
            "form",
            "headers",
            "json",
            "method",
            "origin",
            "url",
          ],
          "type": "object",
        },
        "type": "string",
      },
      "headers": {
        "access-control-allow-credentials": {
          "enum": [
            "true",
            "false",
          ],
          "format": "boolean",
          "type": "string",
        },
        "access-control-allow-origin": {
          "maxLength": 1,
          "minLength": 1,
          "type": "string",
        },
        "connection": {
          "maxLength": 10,
          "minLength": 10,
          "type": "string",
        },
        "content-length": {
          "maxLength": 3,
          "minLength": 3,
          "type": "string",
        },
        "content-type": {
          "maxLength": 16,
          "minLength": 16,
          "type": "string",
        },
        "date": {
          "format": "utc-date-time",
          "type": "string",
        },
        "server": {
          "maxLength": 15,
          "minLength": 15,
          "type": "string",
        },
      },
      "statusCode": 200,
    },
  },
]
`;
